### PR TITLE
feat(rome_js_parser): EcmaScript @decorators class #4252

### PR DESCRIPTION
**Pull request by denbezrukov**
Implement basic decorator cases for classes

## Summary

ECMA spec page of the decorators:
https://arai-a.github.io/ecma262-compare/?pr=2417
https://github.com/tc39/proposal-decorators

This PR adds basic decorators syntax for classes:
```
// class expressions
let a = @decorator class {};
let b = @decorator @functionDecorator(1,2,3) class {};
let c = @first @second class Foo {}

// class declarations
@decorator class Foo {};
@decorator @functionDecorator(1,2,3) class Bar {};
@first @second class Baz {}

// abstract class declarations
@decorator abstract class Foo {};
@decorator @functionDecorator(1,2,3) abstract class Bar {};
@first @second abstract class Baz {}

// exported class declarations
export @decorator class Foo {};
export @decorator @functionDecorator(1,2,3) class Bar {};
export @first @second class Baz {}

```

It's important to mention that `decorator` can be _before_ and _after_ `export` keyword. This PR doesn't implement before syntax:
https://babeljs.io/blog/2023/02/20/7.21.0#decorators-updates
https://devblogs.microsoft.com/typescript/announcing-typescript-5-0-rc/#whats-new-since-the-beta

```
@decorator 
export class Foo {};
```
It's still invalid syntax for Rome. I'm going to implement it in another PR.

## Changelog

- [ ] The PR requires a changelog line

## Documentation

- [ ] The PR requires documentation
- [ ] I will create a new PR to update the documentation

### CODE REVIEW

File: xtask/codegen/js.ungram [hunk 1]
```diff
@@ -578,6 +578,7 @@ JsShorthandPropertyObjectMember = name: JsReferenceIdentifier
 ///////////////
 
 JsClassDeclaration =
+  decorators: JsDecoratorList
```
Comment (ematipico) [inline xtask/codegen/js.ungram:581]: nit: missing tab

File: crates/rome_js_parser/src/syntax/class.rs [hunk 2]
```diff
@@ -141,12 +143,16 @@ pub(super) fn parse_class_expression(p: &mut JsParser) -> ParsedSyntax {
 ///
 /// A class can be invalid if
 /// * It uses an illegal identifier name
-pub(super) fn parse_class_declaration(p: &mut JsParser, context: StatementContext) -> ParsedSyntax {
+pub(super) fn parse_class_declaration(
+    p: &mut JsParser,
+    decorators_list: ParsedSyntax,
```
Comment (ematipico) [inline crates/rome_js_parser/src/syntax/class.rs:148]: nit: `decorator_list`

File: crates/rome_js_parser/src/syntax/class.rs [hunk 3]
```diff
@@ -2377,3 +2395,93 @@ impl ClassMemberModifiers {
...
+}
+
+fn parse_decorator(p: &mut JsParser) -> ParsedSyntax {
+    if !p.at(T![@]) {
+        return Absent;
+    }
+
+    let m = p.start();
+    p.bump(T![@]);
+    parse_lhs_expr(p, ExpressionContext::default().and_in_ts_decorator(true))
```
Comment (ematipico) [inline crates/rome_js_parser/src/syntax/class.rs:2445]: I checked the spec, and not all expressions are allowed.

Screenshot 2023-04-11 at 09 47 11

In `js.ungram` you specified `AnyJsExpression`, but only two expressions are allowed after `@`, which are:
- `JsParenthesizedExpression`
- `JsIdentifierName` + `JsParenthesizedExpression`


Based on the current parsing logic, it seems that any expression is allowed, even examples like
```js
@"expression" + "expression"
class Foo {}
```
I think the previous example should emit errors?

We have a few alternatives:
- we make our AST more strict, where we shape it in a way where we can accept only certain nodes;
- we keep the design of the AST as is, but we still need to flag errors for expressions that are not allowed;
- we do both;

File: crates/rome_js_parser/src/syntax/class.rs [hunk 4]
```diff
@@ -2377,3 +2395,93 @@ impl ClassMemberModifiers {
...
+// // abstract class declarations
+// @decorator abstract class Foo {};
+// @decorator @functionDecorator(1,2,3) abstract class Bar {};
+// @first @second abstract class Baz {}
+// // exported class declarations
+// export @decorator class Foo {};
+// export @decorator @functionDecorator(1,2,3) class Bar {};
+// export @first @second class Baz {}
+
+// test ts decorator_class_not_top_level
```
Comment (ematipico) [inline crates/rome_js_parser/src/syntax/class.rs:2417]: I believe we should add some test cases where we test that **not all expressions** are valid, and we emit errors accordingly

**Review by ematipico — COMMENTED** [review]
Thank you for kicking off this PR! I added some initial feedback :)

File: crates/rome_js_parser/src/syntax/class.rs (hunk 3 above)
Comment (denbezrukov) [reply crates/rome_js_parser/src/syntax/class.rs]: Introduced `AnyJsDecorator`:
```
AnyJsDecorator =
	JsParenthesizedExpression
	| JsCallExpression
	| JsStaticMemberExpression
	| JsIdentifierExpression
	| JsBogusExpression
```

 if an expression has a different kind it sets error `invalid_decorator_error`.

**Comment by ematipico** [comment]
!bench_parser

**Pull request closed without merging**